src/bms/parse/token_processor/bmp.rs

```diff
@@ -60,35 +60,36 @@ impl TokenProcessor for BmpProcessor {
     ) -> core::result::Result<Self::Output, ParseErrorWithRange> {
         let mut objects = BmpObjects::default();
         ctx.all_tokens(|token, prompter| match token.content() {
-            Token::Header { name, args } => {
-                match self.on_header(name.as_ref(), args.as_ref(), prompter, &mut objects) {
-                    Ok(()) => Ok(Vec::new()),
-                    Err(warn) => Ok(vec![warn.into_wrapper(token)]),
-                }
-            }
+            Token::Header { name, args } => Ok(self
+                .on_header(name.as_ref(), args.as_ref(), prompter, &mut objects)
+                .map(|()| Vec::new())
+                .unwrap_or_else(|warn| vec![warn.into_wrapper(token)])),
             Token::Message {
                 track,
                 channel,
                 message,
-            } => {
-                match self.on_message(
+            } => Ok(self
+                .on_message(
                     *track,
                     *channel,
                     message.as_ref().into_wrapper(token),
                     prompter,
                     &mut objects,
-                ) {
-                    Ok(ws) => Ok(ws),
-                    Err(warn) => Ok(vec![warn.into_wrapper(token)]),
-                }
-            }
+                )
+                .unwrap_or_else(|warn| vec![warn.into_wrapper(token)])),
             Token::NotACommand(_) => Ok(Vec::new()),
         })?;
         Ok(objects)
     }
 }
 
 impl BmpProcessor {
+    fn bga_layer(channel: Channel) -> Result<BgaLayer> {
+        BgaLayer::from_channel(channel).ok_or(ParseWarning::SyntaxError(format!(
+            "invalid channel for BgaLayer: {channel:?}",
+        )))
```

Comment:
```suggestion
        BgaLayer::from_channel(channel).ok_or_else(|| {
            ParseWarning::SyntaxError(format!("invalid channel for BgaLayer: {channel:?}",))
        })
```
You should use `ok_or_else` to avoid unnecessary allocation by `format!`.